toolkit/tools/pkg/imagecustomizerlib/artifactsinputoutput.go

```diff
@@ -111,3 +157,157 @@ func outputArtifacts(items []imagecustomizerapi.OutputArtifactsItemType, outputD
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func InjectFiles(buildDir string, baseConfigPath string, inputImageFile string,
+	metadata []imagecustomizerapi.InjectArtifactMetadata,
+) error {
+	logger.Log.Debugf("Injecting Files")
+
+	ic := &ImageCustomizerParameters{}
+	buildDirAbs, err := filepath.Abs(buildDir)
+	if err != nil {
+		return err
+	}
+	ic.inputImageFile = inputImageFile
+	ic.inputImageFormat = strings.TrimLeft(filepath.Ext(ic.inputImageFile), ".")
+	ic.rawImageFile = filepath.Join(buildDirAbs, BaseImageName)
+	ic.outputImageFile = ic.inputImageFile
+	ic.outputImageFormat = imagecustomizerapi.ImageFormatType(ic.inputImageFormat)
+	if err := ic.outputImageFormat.IsValid(); err != nil {
+		return fmt.Errorf("invalid output image format:\n%w", err)
+	}
+	defer func() {
+		cleanupErr := cleanUp(ic)
+		if cleanupErr != nil {
+			if err != nil {
+				err = fmt.Errorf("%w:\nfailed to clean-up:\n%w", err, cleanupErr)
+			} else {
+				err = fmt.Errorf("failed to clean-up:\n%w", cleanupErr)
+			}
+		}
+	}()
+
+	err = convertImageToRaw(ic)
+	if err != nil {
+		return err
+	}
+
+	loopback, err := safeloopback.NewLoopback(ic.rawImageFile)
+	if err != nil {
+		return fmt.Errorf("failed to connect to image file to inject files:\n%w", err)
+	}
+	defer loopback.Close()
+
+	diskPartitions, err := diskutils.GetDiskPartitions(loopback.DevicePath())
+	if err != nil {
+		return err
+	}
+
+	partitionsToMountpoints := make(map[imagecustomizerapi.InjectFilePartition]string)
+	var mountedPartitions []*safemount.Mount
+
+	for idx, item := range metadata {
+		partitionKey := imagecustomizerapi.InjectFilePartition{MountIdType: item.Partition.MountIdType, Id: item.Partition.Id}
+		if _, exists := partitionsToMountpoints[partitionKey]; !exists {
+			partitionsToMountpoints[partitionKey] = filepath.Join(buildDir, fmt.Sprintf("inject-partition-%d", idx))
+
+			partition, _, err := findPartition(item.Partition.MountIdType, item.Partition.Id, diskPartitions, buildDir)
+			if err != nil {
+				return err
+			}
+
+			mount, err := safemount.NewMount(partition.Path, partitionsToMountpoints[partitionKey], partition.FileSystemType, 0, "", true)
+			if err != nil {
+				return fmt.Errorf("failed to mount partition (%s):\n%w", partition.Path, err)
+			}
+
+			logger.Log.Infof("Mounted partition %s at %s", partition.Path, partitionsToMountpoints[partitionKey])
+			mountedPartitions = append(mountedPartitions, mount)
+			defer mount.Close()
+		}
+
+		srcPath := filepath.Join(baseConfigPath, item.Source)
+		destPath := filepath.Join(partitionsToMountpoints[partitionKey], item.Destination)
+		err := file.Copy(srcPath, destPath)
+		if err != nil {
+			return fmt.Errorf("failed to copy binary from (%s) to (%s):\n%w", srcPath, destPath, err)
+		}
+	}
+
+	for _, m := range mountedPartitions {
+		if err := m.CleanClose(); err != nil {
+			logger.Log.Warnf("Failed to cleanly unmount %s: %v", m.Target(), err)
```

Comment:
Return an error if `CleanClose` fails.

The `defer mount.Close()` calls above will ensure the remaining mounts will be closed if this fails.